Office Add-in built on Blazor WASM Web Technologies

aafvstam opened the pull request:
|        Q        |                    A                    |
| --------------- | --------------------------------------- |
| Bug fix?        | no                               |
| New feature?    | yes                               |
| New sample?     | yes                               |
| Related issues? | no                                |

## What's in this Pull Request?

Initial Pull Request to bring in the Office Add-in based on Blazor Webassembly technologies using  Blazor, .NET Core, C# and JavaScript Interop

## Guidance

Just handle as a common OfficeJS Add-in. Blazor WASM will be handled just like any other Add-in on cross-platform environments that are compatible with WASM

davidchesnut commented:
Hi @aafvstam, This is awesome!

Can you please add a copyright to all html, cs, js files? Something similar to follow would work.
```javascript
/*
* Copyright (c) Maarten van Stam
* Licensed under the MIT license.
*/
```

Also can you please add JS doc and C# function summaries to the code?

Thanks!
David

davidchesnut commented:
@aafvstam, for the manifest I think we should just provide the local one for now. I'm looking into options to provide our own "live" host that would allow a developer to sideload the add-in without having to build it. But not sure how long that will take.

aafvstam commented:
Added the copyright headers, added comments, removed the redundant manifest, merged @davidchesnut PR to sideload and fixed some bugs in the main page sample.

Have a look at it and let me know what you think

aafvstam commented:
I just noticed the Read Me file still has info on the double manifest option. This needs to be changed over to the new feature added by @davidchesnut to simply run from Visual Studio on F5.

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.cs:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        /// <summary>
+        /// Invoke JavaScript function to setup the document (insert a couple of paragraphs)
```
```suggestion
        /// Invoke JavaScript function to set up the document (insert a couple of paragraphs)
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.cs:
```diff
@@ -0,0 +1,65 @@
+            await JSModule.InvokeVoidAsync("setupDocument");
+
+        /// <summary>
+        /// Invoke JavaScript function to surround the Paragraphs with Content Controls
```
```suggestion
        /// Invoke JavaScript function to surround the Paragraphs with content controls
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.cs:
```diff
@@ -0,0 +1,65 @@
+            await JSModule.InvokeVoidAsync("insertContentControls");
+
+        /// <summary>
+        /// Invoke JavaScript function to tag the Content Controls with "odd" or "even" tags
```
```suggestion
        /// Invoke JavaScript function to tag the content controls with "odd" or "even" tags
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.cs:
```diff
@@ -0,0 +1,65 @@
+            await JSModule.InvokeVoidAsync("tagContentControls");
+
+        /// <summary>
+        /// Invoke JavaScript function to modify the Content Controls and give some decoration colors and info texts
```
```suggestion
        /// Invoke JavaScript function to modify the content controls and give some decoration colors and info texts
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.cs:
```diff
@@ -0,0 +1,65 @@
+            await JSModule.InvokeVoidAsync("modifyContentControls");
+
+        /// <summary>
+        /// Invoke JavaScript function to register events in case the Content Controls are Deleted or Selection was changed
```
```suggestion
        /// Invoke JavaScript function to register events in case the content controls are deleted or the selection was changed/
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.cs:
```diff
@@ -0,0 +1,65 @@
+            await JSModule.InvokeVoidAsync("registerEvents");
+
+        /// <summary>
+        /// Invoke JavaScript function to revoke registration, remove event handlers from the Content Controls
```
```suggestion
        /// Invoke JavaScript function to revoke registration, remove event handlers from the content controls.
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.cs:
```diff
@@ -0,0 +1,65 @@
+            await JSModule.InvokeVoidAsync("deregisterEvents");
+
+        /// <summary>
+        /// Invoke JavaScript function delete the first 'even' Content Control
```
```suggestion
        /// Invoke JavaScript function delete the first 'even' content control.
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.js:
```diff
@@ -0,0 +1,245 @@
+        context.document.body.insertParagraph("Inserting another paragraph. ", "Start");
+
+        context.document.body.insertParagraph(
+            "Video provides a powerful way to help you prove your point. When you click Online Video, you can paste in the embed code for the video you want to add. You can also type a keyword to search online for the video that best fits your document.",
```
```suggestion
            "Effective add-ins offer unique and compelling functionality that extends Office applications in a visually appealing way. To create a great add-in, provide an engaging first-time experience for your users, design a first-class UI experience, and optimize your add-in's performance.",
```
Replacing with our own text from docs (the previous was from an existing web site. )

aafvstam replied:
Fixed by earlier commit

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.js:
```diff
@@ -0,0 +1,245 @@
+        context.document.body.paragraphs
+            .getLast()
+            .insertText(
+                "To make your document look professionally produced, Word provides header, footer, cover page, and text box designs that complement each other. For example, you can add a matching cover page, header, and sidebar. Click Insert and then choose the elements you want from the different galleries. ",
```
```suggestion
                "With Word add-ins, you can use familiar web technologies such as HTML, CSS, and JavaScript to build a solution that can run in Word across multiple platforms, including on the web, Windows, Mac, and iPad. Learn how to build, test, debug, and publish Word add-ins.",
```

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.js:
```diff
@@ -0,0 +1,245 @@
+ * @param  {} args
+ */
+async function handleContentControlAdded(args) {
+    console.log("Content Control Added!");
```
```suggestion
    console.log("Content control added!");
```

aafvstam replied:
Fixed by earlier commit

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.js:
```diff
@@ -0,0 +1,245 @@
+    console.log("Content Control Added!");
+}
+/**
+ * Step 2: Create Content Controls from the Paragraphs
```
```suggestion
 * Step 2: Create content controls from the paragraphs.
```

aafvstam replied:
Fixed by earlier commit

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.js:
```diff
@@ -0,0 +1,245 @@
+}
+
+/**
+ * ContentControlDeleted Event Handler
```
```suggestion
 * ContentControlDeleted event handler
```

aafvstam replied:
Fixed by earlier commit

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.js:
```diff
@@ -0,0 +1,245 @@
+ * @param  {} args
+ */
+async function handleContentControlDeleted(args) {
+    console.log("Content Control Deleted!");
```
```suggestion
    console.log("Content control deleted!");
```

aafvstam replied:
Fixed by earlier commit

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor.js:
```diff
@@ -0,0 +1,245 @@
+}
+
+/**
+ * SelectionChanged Event Handler
```
```suggestion
 * SelectionChanged event handler
```

aafvstam replied:
Fixed by earlier commit

davidchesnut reviewed (changes requested):
Some minor capitalization and related changes from our style guide.

lindalu-MSFT commented on Samples/word-blazor-add-in/README.md:
```diff
@@ -0,0 +1,118 @@
+
+This sample shows how to build a Word add-in using .NET Blazor technologies. Blazor Webassembly allows you to build your addins using .NET, C# and JavaScript to interact with the OfficeJS API. The add-in uses JavaScript to work with the document and Office JS APIs, but you build the user interface and all other non-Office interactions in C# and .NET Core Blazor technologies.
+
+- Work with Blazor Webassembly in the context of Office.
```
This bulleted list should have a lead-in sentence such as "This sample shows you the following." Or something like that! But it's also sort of redundant to the preceding paragraph so you could delete it.

aafvstam replied:
Fixed in earlier commit

lindalu-MSFT commented on Samples/word-blazor-add-in/README.md:
```diff
@@ -0,0 +1,118 @@
+
+## Run the sample
+
+1. Download or clone this repository.
```
Which repo?

aafvstam replied:
Fixed in upcoming commit

lindalu-MSFT commented on Samples/word-blazor-add-in/README.md:
```diff
@@ -0,0 +1,118 @@
+1. Choose **Debug** > **Start Debugging**. Or press F5 to start the solution.
+1. When Word opens, choose **Home** > **Show Taskpane**.
+
+Now you can try out the controls.
```
```suggestion
Next, try out the controls.
```

lindalu-MSFT commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor.cs:
```diff
@@ -0,0 +1,31 @@
+        }
+
+        /// <summary>
+        /// Basic function to invoke inserting a Paragraph
```
```suggestion
        /// Basic function to invoke inserting a paragraph
```

lindalu-MSFT commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor.js:
```diff
@@ -0,0 +1,17 @@
+﻿/* Copyright(c) Maarten van Stam. All rights reserved. Licensed under the MIT License. */
+/**
+ * Basic function to show how to insert a Paragraph at the start of the Word Document
```
```suggestion
 * Basic function to show how to insert a paragraph at the start of the Word document
```

lindalu-MSFT commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor.js:
```diff
@@ -0,0 +1,17 @@
+
+    return Word.run((context) => {
+
+        // insert a paragraph at the start of the document.
```
```suggestion
        // Inserts a paragraph at the start of the document.
```
We should be consistent with tenses in comments. Also, add a period if comment is a complete sentence. I'm ok with leaving them off sentence fragments.

lindalu-MSFT commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor.js:
```diff
@@ -0,0 +1,17 @@
+        // insert a paragraph at the start of the document.
+        const paragraph = context.document.body.insertParagraph("Hello World from Blazor", Word.InsertLocation.start);
+
+        // sync the context to run the previous API call, and return.
```
```suggestion
        // Sync the context to run the previous API call, and return.
```

lindalu-MSFT commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/InteractWithParagraphs.razor.js:
```diff
@@ -0,0 +1,67 @@
+    });
+}
+/**
+ * Replacing text in the last paragraph found in the document
```
```suggestion
 * Replaces text in the last paragraph found in the document
```

lindalu-MSFT commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/InteractWithParagraphs.razor.js:
```diff
@@ -0,0 +1,67 @@
+    });
+}
+/**
+ * This will count the number of paragraphs in the document
```
```suggestion
 * Counts the number of paragraphs in the document
```

elegault commented:
Awesome stuff @aafvstam!! Hey, I'm following the instructions from the readme in your [forked sample](https://github.com/aafvstam/Office-Add-in-samples/tree/aafvstam-blazoraddin/Samples/word-blazor-add-in), and it's not working. This may be a problem in my environment though, but when I F5 in VS 2022 I get this:

![image](https://user-images.githubusercontent.com/4130365/166336358-1b8a5a44-a8c2-48a5-b357-e393ed2ea4d8.png)

Word does open, but there's this:

![image](https://user-images.githubusercontent.com/4130365/166336393-eb105888-941d-4fc0-b55c-cbe28fad8acb.png)

Nothing in the console logs. But the F5 behaviour seems to switch between the above scenario, or a different one without the debug adapter error and the index.razor page gets loaded in Edge; is that the page that is supposed to be hosted in the task pane? I haven't looked too in depth yet. However, clicking any of the buttons on that page shows the below error at the bottom:

![image](https://user-images.githubusercontent.com/4130365/166336812-d4168fa9-310c-44ee-a552-6f273457ae7f.png)

davidchesnut commented:
@elegault curious if anything in this [stackoverflow article](https://stackoverflow.com/questions/70374907/the-visual-studio-2022-error-failed-to-launch-debug-adapter-additional-informa) fixes the issue you are seeing?

elegault commented on Samples/word-blazor-add-in/word-blazor-sideloader/word-blazor-sideloaderManifest/word-blazor-sideloader.xml:
```diff
@@ -0,0 +1,143 @@
+        <bt:Image id="Contoso.tpicon_80x80" DefaultValue="~remoteAppUrl/Images/Button80x80.png" />
+      </bt:Images>
+      <bt:Urls>
+        <bt:Url id="Contoso.DesktopFunctionFile.Url" DefaultValue="~remoteAppUrl/Functions/FunctionFile.html" />
```
I don't see a Functions/FunctionFile.html folder or file in the project

davidchesnut replied:
fixed.

aafvstam replied:
I think this one is in the 'default' manifest (in case you are using Functions?) @davidchesnut is this a required entry in the manifest? 

(I always have a hard time finding the manifest reference although I have been submitting a few fixes for that months ago)

davidchesnut replied:
I removed it. I think it was from a manifest I copied from another project.

davidchesnut replied:
although the [docs do state it is required](https://docs.microsoft.com/javascript/api/manifest/desktopformfactor). Interesting because I am able to run this manifest without it.

elegault commented:
> @elegault curious if anything in this [stackoverflow article](https://stackoverflow.com/questions/70374907/the-visual-studio-2022-error-failed-to-launch-debug-adapter-additional-informa) fixes the issue you are seeing?

@davidchesnut No, enabling the two "Enable using the multi-target JavaScript debugger..." options has no effect. What I discovered works is changing the startup project for the solution to the word-blazor-sideloader project. However, the add-in is still not loading in Word

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/FirstAddin.razor.css:
```diff
@@ -0,0 +1 @@
+﻿
```
Not really (the FirstAddin files have been renamed to Quotes btw), but I wanted to add Blazor CSS Isolation into the demo:
https://docs.microsoft.com/en-us/aspnet/core/blazor/components/css-isolation?view=aspnetcore-6.0
The file does not have any CSS at this time but we can add this soon (give every quote a different color using this isolated CSS file) so I'd say keep it in for now and 'enhance' the demo a bit later.

elegault commented:
@davidchesnut @aafvstam Regarding the add-in load issue: I still don't see any errors or logging of any kind in Edge DevTools. This is this log file, but I don't see anything obvious.
[visualstudio-js-debugger.txt](https://github.com/OfficeDev/Office-Add-in-samples/files/8654856/visualstudio-js-debugger.txt)

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/FirstAddin.razor.js:
```diff
@@ -0,0 +1,77 @@
+        range.insertText('"Hitch your wagon to a star."\n', Word.InsertLocation.replace);
+
+        // Synchronize the document state by executing the queued commands,
+        // and return a promise to indicate task completion.
```
The comment refers to the line:
`        await context.sync();`
specifically to the sync call on the context **context.sync()**
Technically the function **does** end here and returns the Promise don't you think? (unless an error occurs driving it into the catch errorhandler 😉, but in that case I wonder what happens with the Promise and is it invalidating the comment)

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor:
```diff
@@ -0,0 +1,15 @@
+
+<div class="row">
+	<h1>
+		Call OfficeJS
```
Fixed by your commit, now "Call Office JavaScript API"

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor:
```diff
@@ -0,0 +1,15 @@
+		Call OfficeJS
+	</h1>
+	<p>
+		Write Some Text to the Word Document using JavaScript Interop between Blazor and OfficeJS
```
Fixed by your commit:
Now "Write some text to the Word document using JavaScript interop between Blazor and the Office JavaScript API."

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor.js:
```diff
@@ -0,0 +1,17 @@
+    return Word.run((context) => {
+
+        // insert a paragraph at the start of the document.
+        const paragraph = context.document.body.insertParagraph("Hello World from Page2.razor.js", Word.InsertLocation.start);
```
Fixed by earlier commit
Now "Hello World from Blazor"

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/HelloWorld.razor.js:
```diff
@@ -0,0 +1,17 @@
+        const paragraph = context.document.body.insertParagraph("Hello World from Page2.razor.js", Word.InsertLocation.start);
+
+        // sync the context to run the previous API call, and return.
+        return context.sync();
```
I wonder too 🤨, I randomly picked a few samples from either this Repo and Docs documentation. Apparently some of the demo's are async others are not. We can keep it in to showcase both options but I would argue that Docs should point out what the best practice is, async or not and update all samples accordingly.

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+<section class="setup ms-font-m">
+	<h5>Page Setup</h5>
+	<p>This will add some initial paragraphs to the document.</p>
+	<button class="btn btn-primary" @onclick="Setup">Setup Document</button>
```
Fixed in earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+
+<section class="samples ms-font-m">
+	<div>
+		<h5>Convert to Content Controls</h5>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+<section class="samples ms-font-m">
+	<div>
+		<h5>Convert to Content Controls</h5>
+		<p>This will convert each paragraph surrounded by a Content Control.</p>
```
Fixed in earlier commit, found another type though ... fixed that as well
Now "Surround each paragraph with a content control."

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	<div>
+		<h5>Convert to Content Controls</h5>
+		<p>This will convert each paragraph surrounded by a Content Control.</p>
+		<button id="insert-controls" class="btn btn-primary" @onclick="InsertContentControls">Convert to Content Controls</button>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	</div>
+
+	<div class="mt-3">
+		<h5>Tag Content Controls</h5>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	<div class="mt-3">
+		<h5>Tag Content Controls</h5>
+		<p>Tag content controls, mark as odd and even.</p>
+		<button id="change-controls" class="btn btn-primary" @onclick="TagContentControls">Tag Content Controls</button>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	</div>
+
+	<div class="mt-3">
+		<h5>Modify Content Controls</h5>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	<div class="mt-3">
+		<h5>Modify Content Controls</h5>
+		<p>Modify content control appearance and content.</p>
+		<button id="change-controls" class="btn btn-primary" @onclick="ModifyContentControls">Modify Content Controls</button>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+
+	<div class="mt-3">
+		<h5>Register Content Controls for Events</h5>
+		<p>This will add the OnDelete and onSelectionChanged event handlers to the Content Controls.</p>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	</div>
+
+	<div class="mt-3">
+		<h5>Register Content Controls for Events</h5>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	<div class="mt-3">
+		<h5>Register Content Controls for Events</h5>
+		<p>This will add the OnDelete and onSelectionChanged event handlers to the Content Controls.</p>
+		<button id="change-controls" class="btn btn-primary" @onclick="RegisterContentControlsEvents">Register Event Handlers</button>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	</div>
+
+	<div class="mt-3">
+		<h5>Delete first 'even' tagged Content Control</h5>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+
+	<div class="mt-3">
+		<h5>Delete first 'even' tagged Content Control</h5>
+		<p>This will delete the first Content Control that was tagged 'even'.</p>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	<div class="mt-3">
+		<h5>Delete first 'even' tagged Content Control</h5>
+		<p>This will delete the first Content Control that was tagged 'even'.</p>
+		<button id="change-controls" class="btn btn-primary" @onclick="DeleteContentControl">Delete first 'even' Content Control</button>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+
+	<div class="mt-3">
+		<h5>De-Register Content Controls for Events</h5>
+		<p>This will remove the OnDelete and onSelectionChanged event handlers to the Content Controls.</p>
```
Fixed by earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/Pages/Index.razor:
```diff
@@ -0,0 +1,59 @@
+	</div>
+
+	<div class="mt-3">
+		<h5>De-Register Content Controls for Events</h5>
```
Fixed by your earlier commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+
+* 223 icons designed to be legible down to 8 pixels
+* Super-light SVG files - 61.8 for the entire set 
+* SVG sprite&mdash;the modern replacement for icon fonts
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+
+#### For code samples and everything else you need to get started with Open Iconic, check out our [Icons](http://useiconic.com/open#icons) and [Reference](http://useiconic.com/open#reference) sections.
+
+### General Usage
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+
+To learn more about SVG Sprites, read [Chris Coyier's guide](http://css-tricks.com/svg-sprites-use-better-icon-fonts/).
+
+#### Using Open Iconic's Icon Font...
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
````diff
@@ -0,0 +1,114 @@
+<img src="/open-iconic/svg/icon-name.svg" alt="icon name">
+```
+
+#### Using Open Iconic's SVG Sprite
````
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+#### Using Open Iconic's Icon Font...
+
+
+##### …with Bootstrap
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
````diff
@@ -0,0 +1,114 @@
+<span class="oi oi-icon-name" title="icon name" aria-hidden="true"></span>
+```
+
+##### …with Foundation
````
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
````diff
@@ -0,0 +1,114 @@
+<span class="fi-icon-name" title="icon name" aria-hidden="true"></span>
+```
+
+##### …on its own
````
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+
+## Getting Started
+
+#### For code samples and everything else you need to get started with Open Iconic, check out our [Icons](http://useiconic.com/open#icons) and [Reference](http://useiconic.com/open#reference) sections.
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+
+### General Usage
+
+#### Using Open Iconic's SVGs
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+[Open Iconic v1.1.1](http://useiconic.com/open)
+===========
+
+### Open Iconic is the open source sibling of [Iconic](http://useiconic.com). It is a hyper-legible collection of 223 icons with a tiny footprint&mdash;ready to use with Bootstrap and Foundation. [View the collection](http://useiconic.com/open#icons)
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+[Open Iconic v1.1.1](http://useiconic.com/open)
+===========
+
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+* Webfont stylesheets (including versions for Bootstrap and Foundation) in CSS, LESS, SCSS and Stylus formats
+* PNG and WebP raster images in 8px, 16px, 24px, 32px, 48px and 64px.
+
+
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/css/open-iconic/README.md:
```diff
@@ -0,0 +1,114 @@
+* PNG and WebP raster images in 8px, 16px, 24px, 32px, 48px and 64px.
+
+
+## Getting Started
```
open-iconic no longer used in the Sample

aafvstam commented on Samples/word-blazor-add-in/.gitattributes:
```diff
@@ -0,0 +1,63 @@
+# the diff markers are never inserted). Diff markers may cause the following 
+# file extensions to fail to load in VS. An alternative would be to treat
+# these files as binary and thus will always conflict and require user
+# intervention with every merge. To do so, just uncomment the entries below
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitattributes:
```diff
@@ -0,0 +1,63 @@
+###############################################################################
+# behavior for image files
+#
+# image files are treated as binary by default.
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitattributes:
```diff
@@ -0,0 +1,63 @@
+# 
+# Convert binary document formats to text before diffing them. This feature
+# is only available from the command line. Turn it on by uncommenting the 
+# entries below.
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+## Ignore Visual Studio temporary files, build results, and
+## files generated by popular Visual Studio add-ons.
+##
+## Get latest from https://github.com/github/gitignore/blob/master/VisualStudio.gitignore
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+# Visual Studio 2017 auto generated files
+Generated\ Files/
+
+# MSTest test Results
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+TestResult.xml
+nunit-*.xml
+
+# Build Results of an ATL Project
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+[Rr]eleasePS/
+dlldata.c
+
+# Benchmark Results
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+project.fragment.lock.json
+artifacts/
+
+# ASP.NET Scaffolding
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+*.vspx
+*.sap
+
+# Visual Studio Trace Files
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+# Visual Studio Trace Files
+*.e2e
+
+# TFS 2012 Local Workspace
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+# TeamCity is a build add-in
+_TeamCity*
+
+# DotCover is a Code Coverage Tool
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+# DotCover is a Code Coverage Tool
+*.dotCover
+
+# AxoCover is a Code Coverage Tool
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+*.[Pp]ublish.xml
+*.azurePubxml
+# Note: Comment the next line if you want to checkin your web deploy settings,
+# but database connection strings (with potential passwords) will be unencrypted
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+**/[Pp]ackages/*
+# except build/, which is used as an MSBuild target.
+!**/[Pp]ackages/build/
+# Uncomment if necessary however generally it will be regenerated when needed
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+# (https://github.com/github/gitignore/pull/2483#issue-259490424)
+#*.snk
+
+# Since there are multiple workflows, uncomment next line to ignore bower_components
```
Fixed in upcoming commit

davidchesnut commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/index.html:
```diff
@@ -0,0 +1,61 @@
+        <a class="dismiss">🗙</a>
+    </div>
+
+    <script src="_framework/blazor.webassembly.js"></script>
```
I'm not sure why but this is in a new Blazor project when you create a new project in VS 2022. So I'm guessing this is correct.

aafvstam replied:
Same as with the css the webassembly.js will be used for Isolated JS files
https://docs.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/?view=aspnetcore-6.0#inject-a-script-after-blazor-starts

The convention <addinname>.lib.module.js is used by Blazor to be able to find the preload file.

aafvstam commented on Samples/word-blazor-add-in/.gitignore:
```diff
@@ -0,0 +1,363 @@
+# Click-Once directory
+publish/
+
+# Publish Web Output
```
Fixed in upcoming commit

aafvstam commented on Samples/word-blazor-add-in/word-blazor-add-in/wwwroot/index.html:
```diff
@@ -0,0 +1,61 @@
+    <link href="css/bootstrap/bootstrap.min.css" rel="stylesheet" />
+    <link href="css/app.css" rel="stylesheet" />
+    
+    <link href="BlazorAddIn.styles.css" rel="stylesheet" />
```
True, this is generated by Blazor to facilitate CSS Isolation Bundling (where you put the .css side by side to the .razor file):
https://docs.microsoft.com/en-us/aspnet/core/blazor/components/css-isolation?view=aspnetcore-6.0#css-isolation-bundling

davidchesnut commented:
@elegault does the add-in load on Word on web? Is this just the desktop failing? I'd clear the cache just to be safe for desktop. Thanks!

davidchesnut reviewed (approved):
Looks great!

elegault commented:
> @elegault does the add-in load on Word on web? Is this just the desktop failing? I'd clear the cache just to be safe for desktop. Thanks!

I'm testing on desktop - I'll try clearing the cache. But I can't debug Word Online because when I change the word-blazor-sideloader project's Start Action to Microsoft Edge and run it, VS throws this error: 

> The Site URL property of the project has not been set. In the Properties window, enter the URL for a SharePoint server.

davidchesnut commented:
@elegault Run VS, and once it is running you can manually sideload the manifest to web. Look in Samples\word-blazor-add-in\word-blazor-sideloader\bin\Debug\OfficeAppManifests for the correct manifest.

aafvstam commented:
@elegault Did you check the startup order and dependencies?

![image](https://user-images.githubusercontent.com/5305682/167506261-7df84897-bca3-4768-a0d8-02a7aba789c5.png)

![image](https://user-images.githubusercontent.com/5305682/167506314-7a224a50-7076-4b6e-b495-0b948dcd9676.png)

It should first build and run the BlazorAddin 'server' and then sideload the manifest (hence the dependency)

Maybe your machine is too fast (or slow 😂)  starting one before the other. In that case @davidchesnut suggestion to start the 'server' first manually and sideloading the manifest after that should work.

aafvstam commented:
@elegault BTW, not sure if I mentioned it already (or I was just thinking this over while on holidays) but wanted to mention that indeed if you click the buttons in the browser (vs in the TaskPane) you get an error reload, as the host (Word) is not active in that context.

jameslin0312 commented:
Is it possible to create a custom control on Blazor Technologies like old ActiveX control? The custom control can be inserted, set properties, and saved with Word, Excel or PowerPoint documents.
